File: src/main/java/algo/backtrack/LetterTilePossibilities_1079.java
package algo.backtrack;

import java.util.HashSet;
import java.util.Set;

public class LetterTilePossibilities_1079 {
  public static int numTilePossibilities(String tiles) {
    var chars = tiles.toCharArray();
    var possibilities = new HashSet<String>();
    backtrack(chars, possibilities, "", new boolean[chars.length]);
    return possibilities.size();
  }

  private static void backtrack(char[] chars, Set<String> possibilities, String seq, boolean[] visited) {
    for (var i = 0; i < chars.length; i++) {
      if (!visited[i]) {
        visited[i] = true;
        var newseq = seq + chars[i];
        possibilities.add(newseq);
        backtrack(chars, possibilities, newseq, visited);
        visited[i] = false;
      }
    }
  }

  public static void main(String[] args) {
    System.out.println(numTilePossibilities("AAB"));
  }
}

File: src/main/kotlin/algo/backtrack/LetterTilePossibilities_1079.kt
package algo.backtrack

class KtLetterTilePossibilities_1079 {
    fun numTilePossibilities(tiles: String): Int {
        val possibilities: HashSet<String> = hashSetOf()
        fun backtrack(tiles: String, possibility: String, visited: BooleanArray) {
            if (possibility != "") {
                possibilities.add(possibility)
            }
            for (i in tiles.indices) {
                if (visited[i]) {
                    continue
                }
                visited[i] = true
                val next = possibility + tiles[i]
                backtrack(tiles, next, visited)
                visited[i] = false
            }
        }
        backtrack(tiles, "", BooleanArray(tiles.length))
        return possibilities.size
    }
}

fun main() {
    val tile = KtLetterTilePossibilities_1079()
    println(tile.numTilePossibilities("AAB"))
    println(tile.numTilePossibilities("AAABBC"))
    println(tile.numTilePossibilities("V"))
}